File: 0012-min-stack/solution.go
package main

type MinStack struct {
	stack    []int
	minStack []int
}

func Constructor() MinStack {
	return MinStack{
		stack:    []int{},
		minStack: []int{},
	}
}

func (m *MinStack) Push(value int) {
	m.stack = append(m.stack, value)

	if len(m.minStack) == 0 || value <= m.minStack[len(m.minStack)-1] {
		m.minStack = append(m.minStack, value)
	}
}

func (m *MinStack) Pop() {
	value := m.stack[len(m.stack)-1]
	m.stack = m.stack[:len(m.stack)-1]

	if value == m.minStack[len(m.minStack)-1] {
		m.minStack = m.minStack[:len(m.minStack)-1]
	}
}

func (m *MinStack) Top() int {
	return m.stack[len(m.stack)-1]
}

func (m *MinStack) GetMin() int {
	return m.minStack[len(m.minStack)-1]
}

File: 0012-min-stack/solution.swift
class MinStack {
    
    private var stack: [Int] = []
    private var minStack: [Int] = []
    init() {
        
    }
    
    func push(_ value: Int) {
        stack.append(value)
        
        if minStack.isEmpty || value <= minStack.last! {
            minStack.append(value
        }
    }
    
    func pop() {
        let value = stack.removeLast()
        if value == minStack.last {
            minStack.removeLast()
        }
    }
    
    func top() -> Int {
        stack.last!
    }
    
    func getMin() -> Int {
        return minStack.last!
    }
}


//Your MinStack object will be instantiated and called as such:
let obj = MinStack()
obj.push(value)
obj.pop()
let ret_3: Int = obj.top()
let ret_4: Int = obj.getMin()

